### PR TITLE
Clean up stale extension methods

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Searched that not used in game repo.
These extensions are useful in pre-.NET 4.5 timeline, but totally out of date now. I'm confidential with every replacement.

### CODE REVIEW

File: osu.Framework/Lists/SortedList.cs [hunk 1]
```diff
@@ -128,7 +128,7 @@ public int RemoveAll(Predicate<T> match)
 
         void ICollection<T>.Add(T item) => Add(item);
 
-        public Enumerator GetEnumerator() => new Enumerator(this);
+        public List<T>.Enumerator GetEnumerator() => list.GetEnumerator();
```
Comment (huoyaoyuan) [inline osu.Framework/Lists/SortedList.cs]: Exposing it's actually backed by `List<T>`. I don't think it as a problem.

File: osu.Framework/Extensions/ExtensionMethods.cs [hunk 2]
```diff
@@ -77,17 +80,15 @@ public static int AddInPlace<T>(this List<T> list, T item, IComparer<T> comparer
         /// <returns></returns>
         public static TValue GetOrDefault<TKey, TValue>(this Dictionary<TKey, TValue> dictionary, TKey lookup) => dictionary.TryGetValue(lookup, out TValue outVal) ? outVal : default;
```
Comment (huoyaoyuan) [inline osu.Framework/Extensions/ExtensionMethods.cs]: This one is actually used in Game repo for effectively only 1 place. Should it be also obsoleted?

**Comment by huoyaoyuan** [comment]
Weird test failure. The seems-related `ThrowIfFaulted` is actually never used. And the test always passes at local.

**Comment by bdach** [comment]
Haven't looked in detail, but `SortedList` might well be inapplicable as a replacement to `AddInPlace` here due to a change in semantics (doesn't allow duplicate keys).

File: osu.Framework/Extensions/ExtensionMethods.cs (hunk 2 above)
Comment (bdach) [reply osu.Framework/Extensions/ExtensionMethods.cs]: I'd be in favour of leaving it. A lack of this built-in has frustrated me personally in other projects too many times before.

File: osu.Framework/Extensions/ExtensionMethods.cs [hunk 3]
```diff
@@ -220,6 +221,7 @@ public static string GetDescription(this object value)
             => value.GetType().GetField(value.ToString())
                     .GetCustomAttribute<DescriptionAttribute>()?.Description ?? value.ToString();
 
+        [Obsolete("Use GetAwaiter().GetResult() if you know it's completed.")] // can be removed 20200516
```
Comment (bdach) [inline osu.Framework/Extensions/ExtensionMethods.cs]: Won't `GetAwaiter().GetResult()` block if it isn't completed? The obsoleted extension would just return silently then, wouldn't it?

Comment (huoyaoyuan) [reply osu.Framework/Extensions/ExtensionMethods.cs]: That's right, but I think it's not necessary to provide full alternative for totally unused.

File: osu.Framework/IO/File/FileSafety.cs [hunk 4]
```diff
@@ -276,6 +279,7 @@ public static void RecursiveMove(string oldDirectory, string newDirectory)
 
         public static string GetExtension(string filename) => Path.GetExtension(filename)?.Trim('.').ToLower();
```
Comment (huoyaoyuan) [inline osu.Framework/IO/File/FileSafety.cs]: Should be either obsolete or documented.

File: osu.Framework/Lists/SortedList.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Framework/Lists/SortedList.cs]: The reason this existed is for performance. It seems the JIT does not properly devirtualise this pattern:

```diff
  |      Method |     Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
  |------------ |---------:|----------:|----------:|------:|------:|------:|----------:|
  |     ListFor | 1.712 us | 0.0156 us | 0.0122 us |     - |     - |     - |         - |
- | ListForEach | 1.217 us | 0.0119 us | 0.0106 us |     - |     - |     - |         - |
+ | ListForEach | 2.588 us | 0.0205 us | 0.0182 us |     - |     - |     - |         - |
```

(with these changes)

File: osu.Framework/IO/File/FileSafety.cs [hunk 5]
```diff
@@ -11,6 +11,7 @@ namespace osu.Framework.IO.File
 {
     public static class FileSafety
```
Comment (smoogipoo) [inline osu.Framework/IO/File/FileSafety.cs]: This entire class should be removed.

File: osu.Framework/Extensions/ExtensionMethods.cs [hunk 6]
```diff
@@ -180,14 +179,16 @@ public static T[][] ToJagged<T>(this T[,] rectangular)
...
+        public static string ToResolutionString(this Size size) => $"{size.Width}x{size.Height}";
 
+        [Obsolete("Why not using StreamWriter?")] // can be removed 20200516
         public static void WriteLineExplicit(this Stream s, string str = @"")
         {
             byte[] data = Encoding.UTF8.GetBytes($"{str}\r\n");
             s.Write(data, 0, data.Length);
         }
 
+        [Obsolete("SecureString itself is obsoleted.")] // can be removed 20200513
```
Comment (smoogipoo) [inline osu.Framework/Extensions/ExtensionMethods.cs]: Unnecessary - remove instead.

File: osu.Framework/Extensions/ExtensionMethods.cs [hunk 7]
```diff
@@ -31,6 +31,7 @@ public static class ExtensionMethods
         /// <param name="match">The predicate that needs to be matched.</param>
         /// <param name="startIndex">The index to start conditional search.</param>
         /// <returns>The matched item, or the default value for the type if no item was matched.</returns>
+        [Obsolete("Use LINQ instead.")] // can be removed 20200516
```
Comment (smoogipoo) [inline osu.Framework/Extensions/ExtensionMethods.cs]: Unnecessary - remove instead.

File: osu.Framework/Extensions/ExtensionMethods.cs [hunk 8]
```diff
@@ -77,17 +78,15 @@ public static int AddInPlace<T>(this List<T> list, T item, IComparer<T> comparer
         /// <returns></returns>
         public static TValue GetOrDefault<TKey, TValue>(this Dictionary<TKey, TValue> dictionary, TKey lookup) => dictionary.TryGetValue(lookup, out TValue outVal) ? outVal : default;
 
+        [Obsolete("Do the boundary comparison yourself to help JIT optimizing.")] // can be removed 20200516
```
Comment (smoogipoo) [inline osu.Framework/Extensions/ExtensionMethods.cs]: Unnecessary - remove instead.

File: osu.Framework/Extensions/ExtensionMethods.cs [hunk 9]
```diff
@@ -77,17 +78,15 @@ public static int AddInPlace<T>(this List<T> list, T item, IComparer<T> comparer
...
         public static bool IsValidIndex<T>(this List<T> list, int index) => index >= 0 && index < list.Count;
 
         /// <summary>
         /// Compares every item in list to given list.
         /// </summary>
+        [Obsolete("Use SequenceEqual instead.")] // can be removed 20200516
```
Comment (smoogipoo) [inline osu.Framework/Extensions/ExtensionMethods.cs]: Unnecessary - remove instead.

File: osu.Framework/Extensions/ExtensionMethods.cs [hunk 10]
```diff
@@ -180,14 +179,16 @@ public static T[][] ToJagged<T>(this T[,] rectangular)
         /// <returns>The inverted array. This is always a square array.</returns>
         public static T[][] Invert<T>(this T[][] array) => array.ToRectangular().Invert().ToJagged();
 
-        public static string ToResolutionString(this Size size) => size.Width.ToString() + 'x' + size.Height;
+        public static string ToResolutionString(this Size size) => $"{size.Width}x{size.Height}";
 
+        [Obsolete("Why not using StreamWriter?")] // can be removed 20200516
```
Comment (smoogipoo) [inline osu.Framework/Extensions/ExtensionMethods.cs]: Unnecessary - remove instead.

File: osu.Framework/Lists/SortedList.cs (hunk 1 above)
Comment (huoyaoyuan) [reply osu.Framework/Lists/SortedList.cs]: Oh, I recall that `List.Enumerator` verifies you've not changed the list during each access. That's the key reason making foreach via List slow.

File: osu.Framework/IO/File/FileSafety.cs (hunk 5 above)
Comment (huoyaoyuan) [reply osu.Framework/IO/File/FileSafety.cs]: This class is still used in several places. I intended to obsolete the entire class at first.
`FileDelete`: it moves to cleanup directory when fails to delete. But currently no call to clean the directory. Should be handled later in `Storage`.
`FileNameStrip`: useful, but can be moved into `Storage` class.
`PathStandarise`: used in game side for handling files in a beatmap.
`GetTempPath`: used in game side in `WorkingBeatmap.Save`. Can be replaced easily.

File: osu.Framework/Extensions/ExtensionMethods.cs [hunk 11]
```diff
@@ -284,5 +235,14 @@ public static DisplayIndex GetIndex(this DisplayDevice display)
...
             }
         }
+
+        /// <summary>
+        /// Standardise the path string using '/' as directory separator.
+        /// Useful as output.
+        /// </summary>
+        /// <param name="path">The path string to standardise.</param>
+        /// <returns>The standardised path string.</returns>
+        public static string PathStandardise(this string path)
```
Comment (smoogipoo) [inline osu.Framework/Extensions/ExtensionMethods.cs]: This feels a little weird, maybe it's better as `ToStandardisedPath`

Comment (huoyaoyuan) [reply osu.Framework/Extensions/ExtensionMethods.cs]: Agree. It's extension method instead of static now.

**Pull request merged**